Content.Server/Singularity/EntitySystems/GravityWellSystem.cs

```diff
@@ -61,201 +58,226 @@ public override void Update(float frameTime)
-    /// <param name="xform">The transform of the gravity well to make pulse.</param>
+    /// <param name="uid">The uid of the gravity well to make scan.</param>
+    /// <param name="gravWell">The state of the gravity well to make scan.</param>
+    /// <param name="xform">The transform of the gravity well to make scan.</param>
     private void Update(EntityUid uid, GravityWellComponent? gravWell = null, TransformComponent? xform = null)
     {
         if (Resolve(uid, ref gravWell))
-            Update(uid, _timing.CurTime - gravWell.LastPulseTime, gravWell, xform);
+            Update(uid, _timing.CurTime - gravWell.LastScanTime, gravWell, xform);
     }
 
     /// <summary>
-    /// Makes a gravity well emit a gravitational pulse and puts it on cooldown.
+    /// Updates the set of captured entities for a gravity well.
     /// </summary>
-    /// <param name="uid">The uid of the gravity well to make pulse.</param>
-    /// <param name="gravWell">The state of the gravity well to make pulse.</param>
-    /// <param name="frameTime">The amount to consider as having passed since the last gravitational pulse by the gravity well. Pulse force scales with this.</param>
-    /// <param name="xform">The transform of the gravity well to make pulse.</param>
+    /// <param name="uid">The uid of the gravity well to make scan.</param>
+    /// <param name="gravWell">The state of the gravity well to make scan.</param>
+    /// <param name="frameTime">The amount to consider as having passed since the last gravitational scan by the gravity well. Scan force scales with this.</param>
+    /// <param name="xform">The transform of the gravity well to make scan.</param>
     private void Update(EntityUid uid, TimeSpan frameTime, GravityWellComponent? gravWell = null, TransformComponent? xform = null)
     {
         if(!Resolve(uid, ref gravWell))
             return;
 
-        gravWell.LastPulseTime = _timing.CurTime;
-        gravWell.NextPulseTime = gravWell.LastPulseTime + gravWell.TargetPulsePeriod;
+        gravWell.LastScanTime = _timing.CurTime;
+        gravWell.NextScanTime = gravWell.LastScanTime + gravWell.TargetScanPeriod;
+        gravWell.Captured.Clear();
+
         if (gravWell.MaxRange < 0.0f || !Resolve(uid, ref xform))
             return;
 
-        var scale = (float)frameTime.TotalSeconds;
-        GravPulse(uid, gravWell.MaxRange, gravWell.MinRange, gravWell.BaseRadialAcceleration * scale, gravWell.BaseTangentialAcceleration * scale, xform);
+        foreach(var entity in _lookup.GetEntitiesInRange(xform.MapPosition, gravWell.MaxRange, flags: LookupFlags.Dynamic | LookupFlags.Sundries))
+        {
+            if(!TryComp<PhysicsComponent?>(entity, out var physics)
+            || physics.BodyType == BodyType.Static)
+                continue;
+
+            if(!CanGravPulseAffect(entity))
+                continue;
+            
+            gravWell.Captured.Add(entity);
+        }
     }
```

Comment:
This method is expensive and should be using entityqueries, also you're not querying for static bodies but checking for them? Ideally this should just be applying a force every tick.

Reply:
Ideally yes, unfortunately the physics system clears all forces applied to entities every tick and if I use impulse to scale according to tick time we're back to entities being yanked around.

I could use a component to track entities in range of the singularity, but I was under the impression that the impending Archetypes would make adding/removing components about an order of magnitude more expensive.

Reply:
Moved to using EntityQueries. That should be fast enough.